### ISSUE TITLE
Let `initial=-jnp.inf` by default in `nn.softmax` and `nn.log_softmax`

### ISSUE THREAD

**Issue by carlosgmartin** [google/jax#20626]
Currently, [`nn.softmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softmax.html) and [`nn.log_softmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.log_softmax.html) raise a `ValueError` when `where` is not `None`, if no `initial` is specified:

```python3
from jax import numpy as jnp, nn

x = jnp.linspace(-1, 1, 4)

mask = jnp.array([True, False, True, True])

try:
    print(nn.softmax(x, where=mask))
except ValueError as e:
    print(repr(e))

print(nn.softmax(x, where=mask, initial=-10000))
print(nn.softmax(x, where=mask, initial=-jnp.inf))
```

```
ValueError("reduction operation max does not have an identity, so to use a where mask one has to specify 'initial'")
[0.08208345 0.         0.31139734 0.60651916]
[0.08208345 0.         0.31139734 0.60651916]
```

This error is raised because the implementations of these functions use [`jnp.max`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.max.html) but pass `initial=None` to it.

In this context, the only natural and most useful value for this argument is `-jnp.inf`. See [this issue](https://github.com/google/jax/issues/18661) for further context.

**Proposed fix:** Let `initial=-jnp.inf` by default in `nn.softmax` and `nn.log_softmax` so that they work out-of-the-box. This avoids confusion for users and saves them the trouble of having to manually specify `initial=-jnp.inf` every time they call these functions with a mask.

I can submit a PR for this.

**Comment by jakevdp** [comment]
Sounds good! Feel free to tag me for the PR review. Thanks!

**Comment by carlosgmartin** [comment]
Great! I'll submit it right after #20578 is merged.
